src/com/xilinx/rapidwright/util/RelocateDCPVertically.java

```diff
@@ -0,0 +1,392 @@
+			Site toSite = toTile.getSites()[frTile.getSiteIndex(frSite)];
+			System.out.printf("     to %12s  y_offset %4d : anchor %14s  %14s\n", cell.getSecond(), verticalMoveOffset, toSite, toTile);
+			clearTargetSiteInsts(mod, toSite, top);
+			if (!mod.isValidPlacement(toSite, top)) {
+				System.out.println("Invalid placement.");
+				return false;
+			}
+
+			ModuleInst mi = top.createModuleInst(cell.getFirst(), mod, true);
+			mi.getCellInst().setCellType(mod.getNetlist().getTopCell());
+			mi.place(toSite);
+		}
+		System.out.println("\n");
+		return true;
+	}
+
+
+	/**
+	 * Get the corresponding Site of the given SiteInst if the module is relocated.
+	 *
+	 * @param mod        The module to relocate
+	 * @param inst       The SiteInst to relocate
+	 * @param anchorSite The target location of the module anchor.
+	 * @return The corresponding site or null if none is available.
+	 */
+	private static Site getCorrespondingValidSite(Module mod, SiteInst inst, Site anchorSite) {
+		Site site = inst.getSite();
+		Tile newTile = mod.getCorrespondingTile(site.getTile(), anchorSite.getTile());
+		if (newTile == null) {
+			return null;
+		}
+		Site newSite = site.getCorrespondingSite(inst.getSiteTypeEnum(), newTile);
+		if (newSite == null) {
+			return null;
+		}
+		return newSite;
+	}
+
+
+	/**
+	 * Unplace all cells placed at the proposed existing SiteInts.
+	 *
+	 * @param proposedAnchorSite The proposed new anchor site
+	 * @param design             The design to operate on
+	 * @return False if an invalid Tile or Site is encountered. True otherwise.
+	 */
+	private static boolean clearTargetSiteInsts(Module mod, Site proposedAnchorSite, Design design) {
+
+		for (SiteInst inst : mod.getSiteInsts()) {
+			if (Utils.isLockedSiteType(inst.getSiteTypeEnum())) {
+				continue;
+			}
+
+			Site newSite = getCorrespondingValidSite(mod, inst, proposedAnchorSite);
+
+			SiteInst si = design.getSiteInstFromSite(newSite);
+			if (design != null && si != null) {
+				design.removeSiteInst(si);
+			}
+		}
+
+		return true;
+	}
+
+
+	/**
+	 * Determine if the given net is a clock net.
+	 *
+	 * @param net The net to check
+	 * @return True if the net is a clock net
+	 */
+	public static boolean isClockNet(Net net) {
+		// TODO: rely on attribute instead of name
+		for (SitePinInst sink : net.getSinkPins()) {
+			if (sink.getName().contains("CLK")) return true;
+		}
+		return false;
+	}
+
+
+	/**
+	 * Combine PIPs on the clock nets that were associated with various wires and set it on the top level.
+	 * This is to avoid misintepretation in Vivado.
+	 *
+	 * @param top      The design with black boxes to fill
+	 * @param cellName A black box name
+	 */
+	public static void combinePIPonClockNets(Design top, String cellName) {
+		System.out.println("Combine PIPs on clock nets of " + cellName);
+
+		List<String> clockNets = new ArrayList<>();
+
+		for (EDIFPortInst p : top.getNetlist().getCellInstFromHierName(cellName).getPortInsts()) {
```

Comment:
You use `top.getNetlist()` a lot. Consider caching it as a variable for brevity if not performance?